src/tests/unit_tls.cpp

```diff
@@ -444,183 +501,234 @@ Test::Result test_dtls_handshake(Botan::TLS::Protocol_Version offer_version,
+                  server->send(server_sent);
+                  }
 
-            if(s2c_traffic.size() > 0)
-               {
-               std::vector<byte> input;
-               std::swap(s2c_traffic, input);
+               const bool corrupt_client_data = (r == 3 && rng.next_byte() % 3 <= 1 && rounds < 10);
+               const bool corrupt_server_data = (r == 4 && rng.next_byte() % 3 <= 1 && rounds < 10);
 
-               if(corrupt_client_data)
+               if(c2s_traffic.size() > 0)
                   {
-                  try
+                  /*
+                  * Use this as a temp value to hold the queues as otherwise they
+                  * might end up appending more in response to messages during the
+                  * handshake.
+                  */
+                  std::vector<byte> input;
+                  std::swap(c2s_traffic, input);
+
+                  if(corrupt_server_data)
                      {
-                     input = Test::mutate_vec(input, true);
-                     size_t needed = client.received_data(input.data(), input.size());
-
-                     if(needed > 0 && result.test_lt("Never requesting more than max protocol len", needed, 18*1024))
+                     try
                         {
-                        input.resize(needed);
-                        Test::rng().randomize(input.data(), input.size());
-                        client.received_data(input.data(), input.size());
+                        input = Test::mutate_vec(input, true);
+                        size_t needed = server->received_data(input.data(), input.size());
+
+                        if(needed > 0 && result.test_lt("Never requesting more than max protocol len", needed, 18*1024))
+                           {
+                           input.resize(needed);
+                           Test::rng().randomize(input.data(), input.size());
+                           client->received_data(input.data(), input.size());
+                           }
+                        }
+                     catch(std::exception&)
+                        {
+                        result.test_note("corruption caused server exception");
                         }
                      }
-                  catch(std::exception&)
+                  else
                      {
-                     result.test_note("corruption caused client exception");
+                     try
+                        {
+                        size_t needed = server->received_data(input.data(), input.size());
+                        result.test_eq("full packet received", needed, 0);
+                        }
+                     catch(std::exception& e)
+                        {
+                        result.test_failure("server error", e.what());
+                        }
                      }
+
+                  continue;
                   }
-               else
+
+               if(s2c_traffic.size() > 0)
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Refactor this code to not nest more than 3 if|for|do|while|switch statements. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://sonarqube.com/coding_rules#rule_key=cpp%3AS134)